src/Raven.Server/Rachis/Leader.cs

```diff
@@ -657,186 +660,147 @@ private void CheckPromotables()
-                            {
-                                tasks.Add(state.TaskCompletionSource.Task);
-                            }
-                            else
-                            {
-                                var tcs = new TaskCompletionSource<(long, object)>(TaskCreationOptions.RunContinuationsAsynchronously);
-                                tasks.Add(tcs.Task);
-                                state = new
-                                    CommandState
-                                // we need to add entry inside write tx lock to avoid
-                                // a situation when command will be applied (and state set)
-                                // before it is added to the entries list
-                                {
-                                    CommandIndex = index,
-                                    TaskCompletionSource = tcs,
-                                    ConvertResult = GetConvertResult(cmd.Command),
-                                };
-                                _entries[index] = state;
-                            }
-                        }
-                        context.Transaction.Commit();
-                    }
+        public async Task<(long Index, object Result)> PutAsync(CommandBase command, TimeSpan timeout)
+        {
+            var rachisMergedCommand = new RachisMergedCommand(leader: this, engine: _engine)
+            {
+                Command = command,
+            };
 
-                    if (tasks.Count > 0)
-                        _newEntry.Set();
+            await _engine.TxMerger.Enqueue(rachisMergedCommand); //wait until 'rachisMergedCommand' is executed (until 'rachisMergedCommand.TaskResult' wont be null).
 
-                    for (int i = 0; i < tasks.Count; i++)
-                    {
-                        list[i].TrySetResult(tasks[i]);
-                    }
-                }
-                catch (Exception e)
-                {
-                    if (_running.IsRaised() == false)
-                    {
-                        e = new NotLeadingException(leaderDisposedMessage, e);
-                    }
-                    foreach (var tcs in list)
-                    {
-                        tcs.TrySetException(e);
-                    }
+            var t = rachisMergedCommand.TaskResult;
+            if (await t.WaitWithTimeout(timeout) == false)
+            {
+                GetConvertResult(command)?.AboutToTimeout();
+                throw new TimeoutException($"Waited for {timeout} but the command {command.RaftCommandIndex} was not applied in this time.");
+            }
 
-                    _errorOccurred.TrySetException(e);
-                }
+            try
+            {
+                return t.Result;
+            }
+            catch (AggregateException e) when (e.InnerExceptions.Count == 1)
+            {
+                throw e.InnerExceptions[0];
             }
```

Comment:
maybe use `GetAwaiter.GetResult()` to avoid extracting the inner exception ?

Reply:
Done